### ISSUE TITLE
Suscribe Long Topic Will Cause Broker Deny of Service (from 1.6.0 to 1.6.5)

### ISSUE THREAD

**Issue by RyanShaw-3083** [eclipse/mosquitto#1412]
In code [lib/util_topic.c](https://github.com/eclipse/mosquitto/blob/master/lib/util_topic.c).
```C
int mosquitto_pub_topic_check(const char *str)
{
	int len = 0;
	while(str && str[0]){
		if(str[0] == '+' || str[0] == '#'){
			return MOSQ_ERR_INVAL;
		}
		len++;
		str = &str[1];
	}
	if(len > 65535) return MOSQ_ERR_INVAL;

	return MOSQ_ERR_SUCCESS;
}

int mosquitto_pub_topic_check2(const char *str, size_t len)
{
	size_t i;

	if(len > 65535) return MOSQ_ERR_INVAL;

	for(i=0; i<len; i++){
		if(str[i] == '+' || str[i] == '#'){
			return MOSQ_ERR_INVAL;
		}
	}

	return MOSQ_ERR_SUCCESS;
}
```
The first function is used to implement topic string filtering, causing problems with filtering long strings.

- Testing: 
Ubuntu 16.04 x86_64 / mosquitto 1.6.4 (apt-get.).

- PoC:
By sending 65535 bytes of the same acceptable topic character (`/`), forcing the program to get stuck in the loop, causing SEGV.

Broker will not work properly when subscribers continue to subscribe the malicious topic. 
There are still many broker servers around the world that do not have authentication enabled or under best practices, and this DoS will have a potential (or direct) impact on MQTT services.

**Comment by RyanShaw-3083** [comment]
When testing test.mosquitto.org did not trigger the exception, sorry to bother. I will continue to check the cause of the exception on my server.

**Comment by RyanShaw-3083** [comment]
Still happened! It's actually a horrible bug.

**Comment by ralight** [comment]
Fixed in the commit, I'll have to deal with this properly tomorrow. Thanks for the report.

**Comment by carnil** [comment]
This should be CVE-2019-11779 comparing with https://mosquitto.org/files/cve/2019-11779/

**Comment by robertobarreda** [comment]
When will these new packages be published in the repo `http://repo.mosquitto.org/debian`?

**Comment by karlp** [comment]
@robertobarreda 1.6.7 includes this fix, 1.5.x you can patch yourself.
